File: src/backend/parseMessages.ts
import {io} from "./server";

const emotesURL = '/emotes/';

export function parseMessageForEmotes(message: string,
                                      emoteUrl: string,
                                      emoteMap: Record<string, string>
): string[] {
    const emoteUrls: string[] = [];

    const regex = /:_.*?:/g;
    const matches = message.match(regex);

    if (matches) {
        matches.forEach((emoteText) => {
            if (emoteMap[emoteText]) {
                let cleanedText = emoteText.replace(/[:_]/g, '');
                const newEmoteUrl = emoteUrl + cleanedText;

                console.log(`Emit ${newEmoteUrl}`);
                io.emit('new-emote', {url: newEmoteUrl});

                emoteUrls.push(emoteMap[emoteText]);
            }
        });
    }

    return emoteUrls;
}

function emitEmotes(emoteUrls: string[]) {
    for (let i = 0; i < emoteUrls.length; i++) {
        console.log(`Emit: ${emoteUrls[i]}`);
        io.emit('new-emote', {url: emoteUrls[i]});
    }
}

File: backend/parseMessages.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func ParseMessageForEmotes(message string, emoteUrl string, emoteMap map[string]string) []string {
	emoteUrls := []string{}
	regex := regexp.MustCompile(`:_.*?:`)
	matches := regex.FindAllString(message, -1)

	for _, emoteText := range matches {
		if url, exists := emoteMap[emoteText]; exists {
			cleanedText := strings.ReplaceAll(emoteText, ":", "")
			cleanedText = strings.ReplaceAll(cleanedText, "_", "")
			newEmoteUrl := emoteUrl + cleanedText

			fmt.Printf("Emit %s\n", newEmoteUrl)
			// Note: In Go, you might use a WebSocket library to emit the new emote URL.
			// This placeholder assumes you have a mechanism to handle real-time communication.
			// For example:
			// websocketConn.Emit("new-emote", map[string]string{"url": newEmoteUrl})

			emoteUrls = append(emoteUrls, url)
		}
	}

	return emoteUrls
}
